tycho-extras/tycho-pomless/src/main/java/org/eclipse/tycho/pomless/AbstractTychoMapping.java

```diff
@@ -261,20 +258,48 @@ protected Charset getPrimaryArtifactCharset() {
         return StandardCharsets.UTF_8;
     }
 
-    protected abstract void initModel(Model model, Reader artifactReader, File artifactFile)
-            throws ModelParseException, IOException;
+    protected abstract void initModel(Model model, Reader artifactReader, File artifactFile) throws IOException;
 
     protected static Properties getBuildProperties(File dir) throws IOException {
+        return loadProperties(new File(dir, "build.properties"));
+    }
+
+    static Properties loadProperties(File propertiesPath) throws IOException {
         Properties properties = new Properties();
-        File buildProperties = new File(dir, "build.properties");
-        if (buildProperties.exists()) {
-            try (FileInputStream stream = new FileInputStream(buildProperties)) {
+        if (propertiesPath.isFile()) {
+            try (FileInputStream stream = new FileInputStream(propertiesPath)) {
                 properties.load(stream);
             }
         }
         return properties;
     }
 
+    static Supplier<Properties> getPropertiesSupplier(File propertiesFile) {
+        FutureTask<Properties> t = new FutureTask<>(() -> loadProperties(propertiesFile));
```

Comment:
I haven't seen this pattern of using a FutureTask for "cached singletons" before (in fact, I've probably never used a FutureTask myself yet). Do you typically use that instead of double-checked locking, initialization-on-demand-holder, or just static maps for caches? I'm really curious to learn about that pattern, I definitely don't ask for changes.

Reply:
This is not a pattern I use regularly. I actually came up with this to implement a concurrent lazy computing cache.
It was something like:

```
  private static class LazyConcurrentCache<K, V> {
    private final Map<K, FutureTask<V>> map = new ConcurrentHashMap<>();

    private final Function<K, FutureTask<V>> mappingFunction;

    public LazyConcurrentCache(Function<K, V> function) {
      this.mappingFunction = k -> new FutureTask<>(() -> function.apply(k));
    }

    public V getValue(K key) throws InterruptedException {
      FutureTask<V> task = map.computeIfAbsent(key, mappingFunction);
      task.run();
    private static class LazyConcurrentCache<K, V> {
    private final Map<K, FutureTask<V>> map = new ConcurrentHashMap<>();

    private final Function<K, FutureTask<V>> mappingFunction;

    public LazyConcurrentCache(Function<K, V> function) {
      this.mappingFunction = k -> new FutureTask<>(() -> function.apply(k));
    }

    public V getValue(K key) throws InterruptedException {
      FutureTask<V> task = map.computeIfAbsent(key, mappingFunction);
      try {
        task.run();
        return task.get();
      } catch(InterruptedException e) {
        throw e;
      } catch(ExecutionException e) {
        Throwable cause = e.getCause();
        if(cause instanceof RuntimeException re) {
          throw re;
        }
        throw (Error) cause;
      }
    }
  }    try {
        return task.get();
      } catch(InterruptedException e) {
        throw e;
      } catch(ExecutionException e) {
        Throwable cause = e.getCause();
        if(cause instanceof RuntimeException re) {
          throw re;
        }
        throw (Error) cause;
      }
    }
  }
```
But in this case I mainly thought it would be a convenient replacement for a Mutable, which would avoid having a inital check if the value is present. But when thinking about it again, due to the extra exceptions using a `o.a.commons3.lang.mutable.Mutable` in the end it is simpler to understand and less fancy.

Your suggestion to use it as solution for the double-checked locking idiom seems interesting. Actually it should be a valid solution that does not suffer the problems of the `null-check, syncronized, null-check` approach. But if that is intended to be used more often I would maybe wrap that in a class that handles the different Exception types more gracefully. For example like in the following example.

The nested class can then be used like this
`new LazyValueHolder.LazyValueHolder<>(function).getValue()`
and only throws a `InterruptedException`. But even for that case default could be implemented (like return null, Optional.empty() or just propagate it) to make it more convenient.

```
    public static class LazyValueHolder<V, E extends Exception> {

        public static interface Callable2<V, E extends Exception> extends Callable<V> {
            V call() throws E;
        }

        private final FutureTask<V> task;

        public LazyValueHolder(Callable2<V, E> supplier) {
            this.task = new FutureTask<>(supplier);
        }

        @SuppressWarnings("unchecked")
        V getValue() throws E, InterruptedException {
            try {
                task.run();
                return task.get();
            } catch (InterruptedException e) {
                throw e;
            } catch (Exception e) {
                throw (E) e.getCause();
            }
        }

        public static class Runtime<V> extends LazyValueHolder<V, RuntimeException> {
            public Runtime(Supplier<V> supplier) {
                super(supplier::get);
            }
        }
    }
```

Reply:
I just wanted to note that we here do not need high-performance to the edge parallel processing, I'm even not sure if project models are even read in parallel. So we should not over optimize here without actual proof that it is a performance bottleneck as it makes debugging such code much harder.

Reply:
> I just wanted to note that we here do not need high-performance to the edge parallel processing, I'm even not sure if project models are even read in parallel. So we should not over optimize here without actual proof that it is a performance bottleneck as it makes debugging such code much harder.

Yes sure. The displayed code, was just in response to @Bananeweizen's question and totally unrelated to this change. :)
I just used `FutureTask` for convenience, but changed it to use a `MutableObject` since it turned out to not be that convenient in the end. So this is plain serial programming. The only optimization is that the properties file is only load on the first translated value found.

Reply:
> just used `FutureTask` for convenience

Just a note, `CompleteableFuture` is more powerful than `FutureTask`